# MetaPOAP
the Metagenomic test for Presence Or Absence of Pathways
MetaPOAP is a statistical tool for predicting the likelihood for the presence or absence of metabolic pathways in metagenome-assembled genomes given estimates of completeness and contamination. The background and implementation of MetaPOAP is described in a forthcoming Bioinformatics paper (preprint <a href="https://drive.google.com/file/d/1XzrLkxwmfTOU1HdJ651f0GC2ufnOGmaY/view">manuscript</a>
 and <a href="https://drive.google.com/file/d/1G1C_2eM4oF6XFYnOGv-ZlKBHUI8CF068/view">SI</a>). 
 MetaPOAP is provided as three Python scripts. metapoap_i.py is interactive on the command line, metapoap_cl.py accepts command line arguments, and metapoap_csv.py works with input and output csv files (an example input is provided as test.csv); usage details are provided in the SI of the Bioinformatics paper linked above. 
If you use MetaPOAP, please cite: 
Ward, LM, PM Shih, and WW Fischer. MetaPOAP: Presence or Absence of Metabolic Pathways in Metagenome-Assembled Genomes. Bioinformatics, in press. DOI:10.1093/bioinformatics/bty510
